src/libstore/build/local-derivation-goal.cc

```diff
@@ -1487,8 +1487,17 @@ void LocalDerivationGoal::startDaemon()
 
 void LocalDerivationGoal::stopDaemon()
 {
-    if (daemonSocket && shutdown(daemonSocket.get(), SHUT_RDWR) == -1)
-        throw SysError("shutting down daemon socket");
+    if (daemonSocket && shutdown(daemonSocket.get(), SHUT_RDWR) == -1) {
+        // on macOS if we are just _waiting_ (accept) for a connection
+        // the socket is not considered connected. Hence shutdown will
+        // complain that it's not connected. Thus simply close it. Which
+        // will also exit the accept call.
+        if (errno == ENOTCONN) {
+            daemonSocket.close();
+        } else {
+            throw SysError("shutting down daemon socket");
+        }
```

Comment:
I am tempted to only allow for this with macOS
```suggestion
#if __APPLE__
        // on macOS if we are just _waiting_ (accept) for a connection
        // the socket is not considered connected. Hence shutdown will
        // complain that it's not connected. Thus simply close it. Which
        // will also exit the accept call.
        if (errno == ENOTCONN)
            daemonSocket.close();
        else
#endif
            throw SysError("shutting down daemon socket");
```

Also, wouldd other servers (`src/nix/daemon.cc`, `src/libstore/gc.cc`) also benefit from something like this?

Reply:
@Ericson2314 what about BSD? My understanding is that this is more of a !linux thing.

I would assume so that other servers might benefit from this as well, but I have not looked into the details there.

Reply:
@Ericson2314 thinking even more about this, why would you want to guard this for macOS (and let's assume bsd as well)? The function wants to shutdown a socket. If the error for doing so yields ENOTCONN (Not connected), wouldn't we just want to close it anyway? The inverse is, we try to shut something down, that's not connected, and then raise an exception. Why? Because we were assuming it was connected? In that case this patch is completely wrong to begin with; because we now on macOS do _not_ hold that invariant anymore.

If we assume the socket _must_ be connected (I don't see the rational as to why we should assume it is connected), we need to find a different solution, _and_ figure out why the socket is not connected when we try to shut it down on macOS (and I would assume it's also not connected on linux, but linux doesn't raise the ENOTCONN for accepted connections?

Reply:
I've create a [simple C example](https://gist.github.com/angerman/50ea2fa7d156431a226d599356ba5180) for this behaviour.
Output on Linux:
```
bound
listening...
thread started...
waiting...
accepting socket...
shutting down...
waiting for child
accept: Invalid argument
```

output on macOS
```
bound
listening...
thread started...
waiting...
accepting socket...
shutting down...
shutdown failed: Socket is not connected
accept: Software caused connection abort
waiting for child
```

According to [`man 2 shutdown`](https://man7.org/linux/man-pages/man2/shutdown.2.html), it's actually supposed to return `ENOTCONN` if there is no connection (e.g. the socket in the above example was never connected to begin with. [`man 2 accept`](https://man7.org/linux/man-pages/man2/accept.2.html), gives us:
> [...]It extracts the first connection request on the queue of pending connections for the listening socket, sockfd, creates a new **connected socket**, and returns a new file descriptor referring to that socket.[...]

(emphasis mine)

This leaves me to argue that linux is just more lenient here, and doesn't properly report ENOTCONN, even though the socket is _not_ connected. 

Hence I do not believe this should be guarded behind some platform flag. Should linux ever fix this behaviour (unlikely), the above code would still be correct.

Reply:
I just looked through the code, and I do not see any other code that checks the shutdown return code and raises an exception _if_ that rc is non-zero. All the other cases seem to ignore the rc.

Reply:
Thanks for this thorough analysis @angerman. Could you just update the comment saying something like "posix says..." so it is clear we are doing the right thing, rather than just hacking around some macOS weirdness?

Reply:
@Ericson2314 Done. I'd caution around macOS weirdness assumptions. Unless it's in apple frameworks, it's likely linux weirdness :p

Reply:
For the record, this is what FreeBSD outputs:
```
bound
listening...
thread started...
waiting...
accepting socket...
shutting down...
shutdown failed: Socket is not connected
accept: Software caused connection abort
waiting for child
```